component.js

```diff
@@ -200,33 +206,35 @@ function factory (options) {
       debug = _shouldComponentUpdate.debug(pattern, logFn);
     }
   }
-
-  function createDefaultArguments (displayName, mixins, render) {
-
-    // (render)
-    if (typeof displayName === 'function') {
-      render      = displayName;
-      mixins      = [];
-      displayName = void 0;
+  
+  var NO_RENDER = "Component definition requires `render` function as its argument"
+
+  // [displayName], [mixins = []], render, [decorate]
+  function createDefaultArguments (/* args */) {
+    var args = Array.prototype.slice.call(arguments);
+    var currArg = 0, argCount = args.length;;
+    
+    var displayName, mixins = [], render, decorate;
+    
+    if (currArg == argCount) throw NO_RENDER;
+    if (typeof args[currArg] === 'string' || args[currArg] instanceof String) {
+      displayName = args[currArg++];
     }
-
-    // (mixins, render)
-    if (typeof displayName === 'object' && typeof mixins === 'function') {
-      render      = mixins;
-      mixins      = displayName;
-      displayName = void 0;
+    
+    if (currArg == argCount) throw NO_RENDER;
```

Comment:
I think a lot of the new checks here is less clear than they already are. This change is a result of having the same interface for a decorated component and a non-decorated component. It might not be necessary if we add another factory.